### ISSUE TITLE
feat: Refactor Medical Concept Embedding

### ISSUE THREAD

**Issue by MarIniOnz** [limebit/medmodels#110]
Add Medical concept embedding feature.

Rememeber to add Word2Vec Embedding to propensity score.
`def run(
        self,
        model: str,
        num_neighbors: int,
        distance_metric: str,
        warm_init: Optional[Word2Vec],
        hyperparam: Optional[Dict[str, Any]],
    ) -> pl.DataFrame:
        """
        Runs Propensity Score matching with the chosen classification method.

        The algorithm trains the chosen classification method on the treated and control
        sets. Y_train is constructed as follows: 1 for each entry of the treated and 0
        for the control set. The probability of the class 1 will be assign as a new
        variable "Prop. score" and used for the nearest neighbor matching as the only
        covariate.

        :param model: classification method to be used, default: "logit". Can be
            chosen from ["logit", "dec_tree", "forest"].
        :type model: str, optional
        :param num_neighbors: number of neighbors to be matched per treated subject.
        :type num_neighbors: int, optional
        :param distance_metric: metric to be used for the matching. Defaults to
            "absolute". Can be chosen from ["absolute", "mahalanobis"].
        :type distance_metric: str, optional
        :param warm_init: word2vec model to be used as warm initialization for the
            node2vec model, default: None.
        :type warm_init: Word2Vec, optional
        :param hyperparam: hyperparameters for the classification model, default: None.
        :type hyperparam: dict
        :return: dataframe as a subset of the control set matched for the treated set.
        :rtype: pd.DataFrame
        """
        logging.info("Running the propensity score matching algorithm")
        # We need to convert the data to float, but only the columns that can be
        # converted to float
        data_treated = pl.DataFrame()
        data_control = pl.DataFrame()

        # Convert columns to float where possible
        for col in self.data_treated.columns:
            try:
                data_treated = data_treated.with_columns(
                    self.data_treated[col].cast(pl.Float64, strict=True).alias(col)
                )
                data_control = data_control.with_columns(
                    self.data_control[col].cast(pl.Float64, strict=True).alias(col)
                )
            except (
                pl.exceptions.ComputeError
            ):  # If the column cannot be converted to float
                pass

        # Convert the Polars DataFrames to NumPy arrays
        treated_array = data_treated.to_numpy()
        control_array = data_control.to_numpy()

        # Add the embeddings to the treated and control arrays
        if warm_init is not None:
            logging.info("Adding embeddings to the treated and control arrays")
            embeddings_treated = np.array(
                [warm_init.wv[str(id)] for id in self.data_treated["id"]]
            )
            embeddings_control = np.array(
                [warm_init.wv[str(id)] for id in self.data_control["id"]]
            )
            treated_array = np.concatenate(
                (treated_array, embeddings_treated),
                axis=1,
            )
            control_array = np.concatenate(
                (control_array, embeddings_control),
                axis=1,
            )

        # Train the classification model
        x_train = np.concatenate((treated_array, control_array))
        len_treated, len_control = len(treated_array), len(control_array)
        y_train = np.concatenate((np.ones(len_treated), np.zeros(len_control)))

        treated_prop, control_prop = calculate_propensity(
            x_train=x_train,
            y_train=y_train,
            treated_test=treated_array,
            control_test=control_array,
            hyperparam=hyperparam,
            model=model,
        )

        # Add propensity score to the original data polars dataframes
        self.data_treated = self.data_treated.with_columns(
            pl.Series("prop_score", treated_prop)
        )
        self.data_control = self.data_control.with_columns(
            pl.Series("prop_score", control_prop)
        )

        matched_control = nearest_neighbor(
            self.data_treated,
            self.data_control,
            num_neighbors=num_neighbors,
            metric=distance_metric,
            covariates=["prop_score"],
        )

        return matched_control.drop("prop_score")
`
